src/pytorch_lightning/utilities/data.py

```diff
@@ -321,8 +327,23 @@ def _dataloader_init_kwargs_resolve_sampler(
     fault_tolerant_mode = _FaultTolerantMode.detect_current_mode()
     batch_sampler = getattr(dataloader, "batch_sampler")
     is_predicting = mode == RunningStage.PREDICTING
+
+    if batch_sampler is not None and disallow_batch_sampler:
+        # Check that we don't have a PyTorch default batch sampler that was instantiated in DataLoader __init__
+        if not (
+            type(batch_sampler) is BatchSampler
+            and batch_sampler.sampler == sampler
+            and dataloader.batch_size == batch_sampler.batch_size
+        ):
+            raise MisconfigurationException(
+                "It is not possible to have a batch sampler in your dataloader, when running on multiple IPU devices."
+            )
     # checking the batch sampler type is different than PyTorch default.
-    if batch_sampler is not None and (type(batch_sampler) is not BatchSampler or is_predicting):
+    if (
+        not disallow_batch_sampler
```

Comment:
I think the logic could be refactored like this

```python
    if batch_sampler is not None:
        if disallow_batch_sampler:
            # Check that we don't have a PyTorch default batch sampler that was instantiated in DataLoader __init__
            if not (...):
                ...
        # checking the batch sampler type is different than PyTorch default.
        elif type(batch_sampler) is not BatchSampler or is_predicting:
            ...
``

Reply:
Done